GRB, CRTB, PRTB validation for GH781

Priyashetty17 opened the pull request:
## Issue: https://github.com/rancher/qa-tasks/issues/781
Adding additional steps in the existing RBAC tests to validate the creation/deletion of GRB, CRTB and PRTB
 
## Solution
Adding additional steps in the existing RBAC tests to validate:
- creation of GRB when a user is created.
- creation of CRTB when a user is added to a downstream cluster.
- creation of PRTB when a user is added to a project.
- deletion of CRTB when a user is removed from a downstream cluster.
- deletion of PRTB when a user is removed from a project.
 
Note: The step to validate the deletion of GRB is omitted as user deletion is performed during the cleanup process. 
 
## Engineering Testing
## Manual Testing
N/A

## Automated Testing
Jenkins Run
**Note**: In this PR, we are not introducing new tests but are only adding additional validation steps to the existing "TestRBAC" test within the "TestRBACTestSuite" test suite. Test "TestRBAC" with the new steps has been run to validate the code works.

## QA Testing Considerations
N/A

## Regressions Considerations

slickwarren reviewed (commented):
can you add a readme for how users should be running this? Also, can you rebase this?

anupama2501 commented on tests/v2/validation/rbac/rbac.go:
```diff
@@ -237,6 +238,19 @@ func createRole(client *rancher.Client, context string, roleName string, rules [
 
 }
 
+func CheckRBACResourceExistsForUser(client *rancher.Client, collection *v1.SteveCollection, userName string) (bool, error) {
```
Can we make this an extension? If the rbac tests are moved to integration suite, we won't be able to access the helper and adding an extension would help us not change anything..

Priyashetty17 replied:
Moved this function to rbac_test.go instead based on your review comment below

anupama2501 commented on tests/v2/validation/rbac/rbac.go:
```diff
@@ -237,6 +238,19 @@ func createRole(client *rancher.Client, context string, roleName string, rules [
+	if err != nil {
+		return false, err
+	}
+	for _, item := range collection.Data {
```
May be we can just move this part to rbac_test.go and add this as a private function? thoughts?

Priyashetty17 replied:
Agree. For now, we can keep this as a private function since this is relevant to the rbac test. Moved this function to rbac_test.go

anupama2501 commented on tests/v2/validation/rbac/rbac_test.go:
```diff
@@ -284,6 +289,12 @@ func (rb *RBTestSuite) ValidateRemoveProjectRoles() {
+	require.NoError(rb.T(), err)
+	_, err = CheckRBACResourceExistsForUser(rb.client, prtbs, rb.standardUser.Username)
+	require.Error(rb.T(), err)
+	log.Info(err.Error())
```
Don't have a strong opinion here but we don't have to log the error here.

Priyashetty17 replied:
Removed

anupama2501 reviewed (commented):
Have a few comments, looks good otherwise. Can you also squash the commits?

Priyashetty17 commented:
@slickwarren I have rebased the code. Addressing your feedback on the readme, I'd like to clarify that in this PR, we are not introducing new tests, We are only adding extra validation steps to the existing "TestRBAC" test within the "TestRBACTestSuite" test suite. To validate the new steps, "TestRBAC" test is what we need to run. I added this as a note under "Automated Testing" section.

slickwarren commented on tests/v2/validation/rbac/rbac_test.go:
```diff
@@ -330,9 +358,21 @@ func (rb *RBTestSuite) TestRBAC() {
 					err := users.AddProjectMember(rb.client, rb.adminProject, rb.standardUser, tt.role)
 					require.NoError(rb.T(), err)
+
+					log.Info("Validating PRTB is created for the user.")
```
I have to read through the code to know what the abbreviations stand for GRB, CRTB and PRTB. Since they are used interchangeably in the logs as-is, can we stick to only using the full name for clarity?

Priyashetty17 replied:
Replaced the abbreviations "GRB, CRTB and PRTB" with their full names.

slickwarren commented on tests/v2/validation/rbac/rbac_test.go:
```diff
@@ -59,6 +60,19 @@ func (rb *RBTestSuite) SetupSuite() {
 
 }
 
+func (rb *RBTestSuite) CheckRBACResourceExistsForUser(client *rancher.Client, collection *v1.SteveCollection, userName string) (bool, error) {
```
this is a public function, will it be used elsewhere? either way, can you move it to rbac.go?

igomez06 replied:
agreed

anupama2501 replied:
This was previously in rbac.go but I wanted it to be in the rbac_test.go as when the rbac_test.go moves to integration, we do not want to replicate rbac.go to integration test and we want the helper functions that are needed to be extensions, so its available for both integration and validation tests. But since this is a special case, this was added in rbac_test.go.
but @Priyashetty17 can you make this a private function instead? 

cc @igomez06

slickwarren commented on tests/v2/validation/rbac/rbac_test.go:
```diff
@@ -330,9 +358,21 @@ func (rb *RBTestSuite) TestRBAC() {
 					err := users.AddClusterRoleToUser(rb.client, rb.cluster, rb.standardUser, tt.role)
 					require.NoError(rb.T(), err)
+
+					log.Info("Validating CRTB is created for the user.")
```
same comment about abbreviations here

slickwarren reviewed (commented):
some nits about placement/naming

anupama2501 commented on tests/v2/validation/rbac/rbac_test.go:
```diff
@@ -59,6 +60,19 @@ func (rb *RBTestSuite) SetupSuite() {
 
 }
 
+func checkRBACResourceExistsForUser(client *rancher.Client, collection *v1.SteveCollection, userName string) (bool, error) {
```
Do we want the boolean to be returned here as we are not using it anywhere else in the called function?

Priyashetty17 replied:
Good catch, we don't need it. Removed it

anupama2501 reviewed (commented):
Have one question, rest all looks great.

anupama2501 reviewed (dismissed):
lgtm

igomez06 commented on tests/v2/validation/rbac/rbac_test.go:
```diff
@@ -59,6 +60,19 @@ func (rb *RBTestSuite) SetupSuite() {
+			return nil
+		}
+	}
+	return errors.New(collection.Data[0].JSONResp["kind"].(string) + " does not exist for user " + userName + "!")
```
This seems like a possible flaky scenario. Why index from zero to get the kind? I think we can make this a little bit more efficient in you list call you should use the list query instead iterating through the whole slice: https://github.com/rancher/rancher/blob/5e1ee26116162ab77998f11e0d53d19663f304c7/tests/v2/integration/steveapi/steve_api_test.go#L2414-L2415

anupama2501 replied:
Good point. This reminded me, we have a similar querying done by Michael to verify the roll out. we can probably use this function and then request for Project or Cluster Role template bindings, code gets the list of RoleBindings. https://github.com/rancher/rancher/blob/release/v2.8/tests/framework/extensions/users/users.go#L289

igomez06 replied:
Yes very similar. Do you mean use the function directly? So export it?

igomez06 replied:
this is a very good point. @anupama2501 ?

anupama2501 replied:
Went through the code and it solves the problem of checking the CRTB/PRTB when a user is created/deleted. So, yes we can mark the ticket as resolved for CRTB/PRTB but for grbs we still need the changes.

Priyashetty17 replied:
Removed CRTB/PRTB related changes. Modified the code to list GRB for a user by using list query as suggested by @igomez06

igomez06 commented on tests/v2/validation/rbac/rbac_test.go:
```diff
@@ -269,16 +282,26 @@ func (rb *RBTestSuite) ValidateRemoveClusterRoles() {
 	require.NoError(rb.T(), err)
+
+	log.Info("Validating Cluster Role Template Binding is deleted for the user.")
+	crtbs, err := rb.client.Steve.SteveType("management.cattle.io.clusterroletemplatebinding").List(nil)
```
This list call can use the query, to get a specific role, and if it's not there, then report a failures.

igomez06 reviewed (commented):
resource check can be much more efficient. That is my only concern with this PR.

igomez06 reviewed (approved):
LGTM

brudnak reviewed (approved):
lgtm!
